### PR TITLE
reorganize and update documentation

### PR DESCRIPTION
**Pull request by johnnychen94**
The main change of this PR is to reorganize the documentation by splitting the manual into two parts: tutorials and package documentations. Files and assets are moved around and renamed because of this.

Several other changes:

* [index.md] replace documentation TOC with packages list
* [install.md]
  * advertise `mosaicview` as recommended method to display multiple images
  * remove the outdated troubleshooting page as image IO and display are no longer an issue.
* [api_comparison.md] [function_reference.md]
  * update functions provided by ImageDistances and ImageQualityIndexes
  * rewrite "Image loading and saving" section
  * add TOC in function references
* [demos] fixes warnings when image value range exceeds [0, 1]
* remove the duplicated documentation to [ImageFiltering.jl](https://juliaimages.org/ImageFiltering.jl/stable/)

Preview: https://juliaimages.org/previews/PR125

### CODE REVIEW

File: docs/src/index.md [hunk 1]
```diff
@@ -26,6 +26,35 @@ in Julia.
     Your changes will be vetted by developers before becoming permanent, so don't
     worry about whether you might say something wrong.
 
-```@contents
-Pages = ["install.md", "quickstart.md", "arrays_colors.md", "conversions_views.md", "indexing.md", "imageaxes.md", "imagefiltering.md", "imagemetadata.md", "imagesegmentation.md", "function_reference.md", "api_comparison.md"]
-```
+The following is a list of packages that are closely related to the JuliaImages ecosystem and also maintained by members of
```
Comment (timholy) [inline docs/src/index.md:29]: I like the idea here, but I worry this list will be intimidating for newcomers ("Welcome to JuliaImages! Here are 60 different packages you have to master. Good luck!"), especially if it appears on the front page. I'd rather start with a gentle introduction before getting into a long list spelling out what code goes where. Let's just do `using Images` until readers gain a bit of experience with JuliaImages.

File: docs/examples/spatial_transformation/alpha_compositing.jl [hunk 2]
```diff
@@ -100,4 +100,4 @@ color_panel = color_panel[axes(red_c)...] # crop empty region
 
 # --- save covers --- #src
 using FileIO #src
-save("assets/alpha_compositing.png", color_panel) #src
+save("assets/alpha_compositing.png", map(clamp01nan, color_panel)) #src
```
Comment (timholy) [inline docs/examples/spatial_transformation/alpha_compositing.jl:103]: :+1:

File: docs/src/api_comparison.md [hunk 3]
```diff
@@ -11,96 +11,98 @@ welcome.
 | Operation                             | JuliaImages                               | scikit-image + NumPy              | Matlab (ImageProcessing + ComputerVision) |
 |---------------------------------------|-------------------------------------------|-----------------------------------|-------------------------------------------|
 | **Input/output**                      |                                           |                                   |                                           |
-| Read image file                       | `load` (FileIO.jl)                        | `imread`                          | `imread`                                  |
-| Write image file                      | `save` (FileIO.jl)                        | `imsave`                          | `imwrite`                                 |
+| Read image file                       | [`load`](@ref)                            | `imread`                          | `imread`                                  |
```
Comment (timholy) [inline docs/src/api_comparison.md:14]: Having these link to the reference page will be a huge usability improvement! :fireworks:

File: docs/src/tutorials/install.md [hunk 4]
```diff
@@ -90,8 +90,27 @@ Currently there're five julia packages can be used to display an image:
...
+```@example mosaicview
+img1 = testimage("mandrill") # 512*512 RGB image
+img2 = testimage("blobs") # 254*256 Gray image
+mosaicview(img1, img2; nrow=1)
+```
+
+```@example mosaicview
+img = testimage("mri-stack") # 226×186×27 Gray image
+mosaicview(img; fillvalue=0.5, npad=2, ncol=7, rowmajor=true)
+```
```
Comment (timholy) [inline docs/src/tutorials/install.md:109]: :+1:

File: docs/src/tutorials/install.md [hunk 5]
```diff
@@ -90,8 +90,27 @@ Currently there're five julia packages can be used to display an image:
...
+
+
 ## Troubleshooting
 
 Reading and writing images, as well as graphical display, involve interactions with external software libraries;
-occasionally, the installation of these libraries goes badly.
-If you experience any difficulties with any of the above steps, please see the [Installation troubleshooting](@ref) page for more information.
+occasionally, the installation of these libraries goes badly. Fortunately, the [artifact system](https://julialang.org/blog/2019/11/artifacts/) shipped with Julia at least 1.3 has made this process much reliable, so if you're experiencing any installation
```
Comment (timholy) [inline docs/src/tutorials/install.md:115]:
```suggestion
occasionally, the installation of these libraries goes badly. Fortunately, the [artifact system](https://julialang.org/blog/2019/11/artifacts/) shipped with Julia at least 1.3 has made this process much more reliable, so if you're experiencing any installation
```

File: docs/src/tutorials/install.md [hunk 6]
```diff
@@ -90,8 +90,27 @@ Currently there're five julia packages can be used to display an image:
...
+trouble, please try with a newer Julia release.
```
Comment (timholy) [inline docs/src/tutorials/install.md:116]:
```suggestion
trouble, please try with Julia 1.3 or higher.
```

**Comment by johnnychen94** [comment]
> Why delete ImageFiltering but keep ImageAxes, ImageFeatures, etc?

Good question. I wrote something but it's incomplete. Here's the idea: I'm reading the JuliaImages doc as:

* tutorials
* packages documentation
* demos
* faqs
* developer guide

Packages documentation comes from two sources: packages that don't have independent docs(e.g., ImageDistances) and that do(e.g., ImageFiltering). I believe it's programmatically doable to collect them together and build a centralized version of docs. For example:

```bash
git clone https://github.com/JuliaImages/ImageFiltering.jl
mv ImageFiltering.jl/docs/ docs/src/packages/ImageFiltering.jl
# modify make.jl to insert docs/src/packages/ImageFiltering.jl into the manual
```

so that we can build the docs based on each package's docs instead of plain links.

**Comment by timholy** [comment]
If we host them here, I agree that building the package's own docs seems like a much better solution than "forking" the docs as is done currently. Alternatively we could make sure each package has its own docs and link those. I am not sure which is better.

WRT things missing from juliaimages' docs, Chris Foster recently pointed me to a lovely resource which you may have already seen, but if not: https://documentation.divio.com/

**Comment by johnnychen94** [comment]
advantages of centralized docs:

* fewer cross-site links so it's easier to maintain references with `@ref`
* better searching function

**Comment by timholy** [comment]
I'm sold! My guess is we can grab the docs without needing to clone the whole repository, similar to how `Pkg.add` gives you "just the code" whereas `Pkg.dev` gives you the entire git history, etc.

**Comment by kimikage** [comment]
I may have to submit a separate issue, but:
> Why don't JuliaImages use the dynamic generation? (https://juliaimages.org/latest/conversions_views/#Using-colorview-to-make-color-overlays-1) :smile:
>
> **Edit:** You’ve probably noticed, but I just meant that the `b` was broken in https://github.com/JuliaImages/juliaimages.github.io/pull/29

_Originally posted by @kimikage in https://github.com/JuliaGraphics/Colors.jl/issues/398#issuecomment-590335795_

**Comment by johnnychen94** [comment]
I guess it's because these pages were initially written in days when there isn't such a feature. I'm gradually switching to that these days; e.g., the quickstart and demos are done in that way.

**Comment by kimikage** [comment]
> I just meant that the `b` was broken in #29

:sweat_smile:

File: docs/src/index.md (hunk 1 above)
Comment (johnnychen94) [reply docs/src/index.md]: This is moved to the front page of the "Pakcages" part. And I've also added some words to explain how the documentation is organized and built.

File: docs/make.jl [hunk 7]
```diff
@@ -13,27 +13,30 @@ format = Documenter.HTML(edit_link = "source",
...
+                         joinpath("tutorials", "quickstart.md"),
+                         joinpath("tutorials", "arrays_colors.md"),
+                         joinpath("tutorials", "conversions_views.md"),
+                         joinpath("tutorials", "indexing.md")
+                     ],
+                     "Packages" => Any[
+                         "Introduction" => joinpath("pkgs", "index.md"),
+                         "ImageAxes.jl" => joinpath("pkgs", "axes", "index.md"),
+                         "ImageMetaData.jl" => joinpath("pkgs", "metadata", "index.md"),
+                         "ImageSegmentation.jl" => joinpath("pkgs", "segmentation", "index.md"),
```
Comment (johnnychen94) [inline docs/make.jl:32]: As you might have noticed, this nested folder structure contributes to a very long URL.

File: docs/make.jl [hunk 8]
```diff
@@ -13,27 +13,30 @@ format = Documenter.HTML(edit_link = "source",
...
-                         "imageaxes.md",
-                         "imagefiltering.md",
-                         "imagemetadata.md",
-                         "imagesegmentation.md",
-                         "imagetransformations.md",
-                         "imagefeatures.md",
-                         "troubleshooting.md",
+                     "Tutorials" => Any[
+                         joinpath("tutorials", "install.md"),
+                         joinpath("tutorials", "quickstart.md"),
```
Comment (johnnychen94) [inline docs/make.jl:23]: Should `install.md` and `quickstart.md` included into `Tutorials`? If the answer is yes, we might need to move https://github.com/JuliaImages/juliaimages.github.io/blob/source/docs/src/quickstart.md#L302-L342 to a better place.

File: docs/src/index.md [hunk 9]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+helps us improve the documentation.
+
+```@setup versions
+using InteractiveUtils
+```
+```@repl versions
+using Pkg, Dates
+today()
+versioninfo()
+Pkg.status()
```
Comment (kimikage) [inline docs/src/index.md:53]: Is this really what we need for the top page? I think the top page should be a "welcome" page, as @timholy implies in https://github.com/JuliaImages/juliaimages.github.io/pull/125#discussion_r412076360.

Comment (kimikage) [reply docs/src/index.md:53]: Another option is using

    ```@raw html
    <details>
      <summary>Package Status</summary>
    ```
and

    ```@raw html
    </details>
    ```
I don't really recommend it, though,

File: docs/src/index.md [hunk 10]
```diff
@@ -26,6 +26,29 @@ in Julia.
     Your changes will be vetted by developers before becoming permanent, so don't
     worry about whether you might say something wrong.
 
-```@contents
-Pages = ["install.md", "quickstart.md", "arrays_colors.md", "conversions_views.md", "indexing.md", "imageaxes.md", "imagefiltering.md", "imagemetadata.md", "imagesegmentation.md", "function_reference.md", "api_comparison.md"]
+This documentation is a collection of two parts: a top-level documentation that guides you into the
```
Comment (timholy) [inline docs/src/index.md:29]: Would it be better to say "This documentation is a collection of several parts." and then delete the next couple of sentences, and launch straight into 'The "Tutorials" part...'?

File: docs/src/index.md [hunk 11]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+JuliaImages ecosystem, and a list of more specific documentation of sub-packages that gives you more
+information of each image processing subfields.
+
+The "Tutorials" part contains a list of tutorials that help you gain better understanding of the JuliaImages
+ecosystems. The ["Packages"](@ref page_packages_index) part contains all related sub-packages in JuliaImages,
```
Comment (timholy) [inline docs/src/index.md:34]:
```suggestion
ecosystem. The ["Packages"](@ref page_packages_index) part contains information about specific components (themselves Julia packages) that together comprise JuliaImages
```

File: docs/src/index.md [hunk 12]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+users can get more information on the specific image processing subfields by reading them. The "Demos"
```
Comment (timholy) [inline docs/src/index.md:35]:
```suggestion
and address specific subfields of image processing. The "Demos"
```

File: docs/src/index.md [hunk 13]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+part gives you a task-by-task introduction of how you could do your daily work with JuliaImages.
```
Comment (timholy) [inline docs/src/index.md:36]:
```suggestion
part gives you demonstrations of how to carry out specific tasks with JuliaImages.
```

File: docs/src/index.md [hunk 14]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+The "References" part is a collection of function references, it's a dictionary of functions in JuliaImages
```
Comment (timholy) [inline docs/src/index.md:37]:
```suggestion
The "References" part is a collection of function references. ```

File: docs/src/index.md [hunk 15]
```diff
@@ -26,6 +26,29 @@ in Julia.
...
+and the recommended way to use it is by the searching function of your browser `Ctrl-F`/`Cmd-F`.
```
Comment (timholy) [inline docs/src/index.md:38]:
```suggestion
The recommended way to use it is by the searching function of your browser `Ctrl-F`/`Cmd-F`.
```

File: docs/src/pkgs/index.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
+# [Packages under JuliaImages](@id page_packages_index)
+
+!!! tip
+    It's recommended to gone through the tutorials before you dive into the package documentation to get
```
Comment (timholy) [inline docs/src/pkgs/index.md:4]:
```suggestion
    Most users should probably start with the tutorials before diving into the documentation for individual packages.
```

File: docs/src/pkgs/index.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    a basic understanding of the JuliaImages ecosystem. Otherwise, you might find yourself lost in the
```
Comment (timholy) [inline docs/src/pkgs/index.md:5]:
```suggestion
    Much of JuliaImages' functionality comes from composing very basic operations that are not always available in other environments,
```

File: docs/src/pkgs/index.md [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    details and get frustrated.
```
Comment (timholy) [inline docs/src/pkgs/index.md:6]:
```suggestion
    and the tutorials will make this workflow clearer.
```

File: docs/src/pkgs/index.md [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+🚧 This section consists of documentations of packages that are closely related to JuliaImages ecosystem.
```
Comment (timholy) [inline docs/src/pkgs/index.md:8]:
```suggestion
🚧 This section documents the individual components that form the JuliaImages ecosystem.
```

File: docs/src/pkgs/index.md [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+Most of them are also maintained by members of JuliaImages. They're ordered alphabetically so you can
```
Comment (timholy) [inline docs/src/pkgs/index.md:9]:
```suggestion
The ones marked with a `*` are available via `using Images`, but you can also use packages individually. Below, they're grouped into broad categories, then ordered alphabetically so you can
```
I wondered if explaining the star at the top would be more helpful than at the bottom, but I can go either way, your choice.

Comment (timholy) [reply docs/src/pkgs/index.md:9]: These provide links to the packages themselves, which seems useful, but should we say something about aggregating the documentation here?

File: docs/src/pkgs/index.md [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    For package developers, `Images.jl` is usually a large dependency to be included in the `deps` section of
+    your `Project.toml`. Hence it is reccomended to add only `ImageCore` together with the exact sub-packages
+    you need. You can use `@which` to find out the exact package and file a method/function belongs to.
+
+JuliaImages is not a closed ecosystem; it works nicely with many other packages outside of JuliaImages.
+The following is an incomplete list of third-party packages that are widely used together with `Images.jl`:
+
+* [Augmentor.jl](https://github.com/Evizero/Augmentor.jl) provides several basic image augmentation
+  operations for image-related machine learning tasks.
+* [Flux.jl](https://github.com/FluxML/Flux.jl) is a featured deep learning toolbox in Julia.
```
Comment (timholy) [inline docs/src/pkgs/index.md:68]:
```suggestion
* [Flux.jl](https://github.com/FluxML/Flux.jl) is a deep learning toolbox in Julia.
```
I'm not sure how to draw the boundary between "featured" and non-featured.

File: docs/src/pkgs/index.md [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    and characterizing "keypoints" (salient features) in images.
+  * [ImageQualityIndexes.jl](https://github.com/JuliaImages/ImageQualityIndexes.jl)\* provides several
+    image quality assessment indexes, e.g., PSNR and SSIM.
+  * [ImageTransformations.jl](https://github.com/JuliaImages/ImageTransformations.jl)\* provides functions
+    related to geometric transformations.
+  * [ImageSegmentation.jl](https://github.com/JuliaImages/ImageSegmentation.jl) provides several image
+    segmentation algorithms.
+
+Items marked with \* means that they are reexported by `Images.jl` via
+[Reexport.jl](https://github.com/simonster/Reexport.jl) so that you can import all of them with only `using Images`.
```
Comment (timholy) [inline docs/src/pkgs/index.md:56]: If you decide to move this above it doesn't need to be repeated here.

File: docs/src/index.md (hunk 9 above)
Comment (johnnychen94) [reply docs/src/index.md]: I believe it is necessary to share the build information somewhere in the documentation. I've now moved this to the troubleshooting section in `install.md`.

> we can find the date and the version from the top-right icon.

Yes but I feel putting them together more convenient to get relevant information.

File: docs/make.jl (hunk 7 above)
Comment (kimikage) [reply docs/make.jl:32]: As this PR shows, modularization makes it easier to access their own assets from "*.md". On the other hand, it makes it difficult to access the common assets, and `@example` and so on run in the directory "build". So I think the length of the local relative path is more important than the URL.
There is not much difference, though.

Comment (kimikage) [reply docs/make.jl:32]: BTW, the assets in "src" will be copied into the "build" before running doctests. So I don't think the "src/" is needed when loading an asset. That improves the consistency between `load` and `save`.

**Comment by johnnychen94** [comment]
This PR is growing too big to review and discuss. Since it's basically becoming stable, I'm merging it.

I opened two issues #127 and #128 for unsolved comments. If there's something I've missed, please feel free to open more issues.

**Comment by kimikage** [comment]
:tada:

BTW, did you read the error log?

**Comment by johnnychen94** [comment]
My bad, I'll update it later.

**Pull request merged**